libdnf/rpm/solv/map.hpp

```diff
@@ -0,0 +1,297 @@
+public:
+    using iterator = SolvMapIterator;
+    iterator begin() const;
+    iterator end() const;
+
+    /// Initialize an empty Map of given size
+    explicit SolvMap(int size);
+
+    /// Clone from an existing Map
+    explicit SolvMap(const Map * other);
+
+    /// Copy constructor: clone from an existing SolvMap
+    SolvMap(const SolvMap & other);
+
+    ~SolvMap();
+
+    // GENERIC OPERATIONS
+
+    /// Return the underlying libsolv Map
+    ///
+    /// @replaces libdnf:sack/packageset.hpp:method:PackageSet.getMap()
+    const Map * get_map() const;
+
+    /// Return the number of solvables in the SolvMap (number of 1s in the bitmap).
+    ///
+    /// @replaces libdnf:sack/packageset.hpp:method:PackageSet.size()
+    std::size_t size() const;
+
+    bool empty() const;
+
+    void clear();
+
+    // ITEM OPERATIONS
+
+    /// @replaces libdnf:sack/packageset.hpp:method:PackageSet.set(Id id)
+    void add(Id id);
+
+    /// @replaces libdnf:sack/packageset.hpp:method:PackageSet.has(Id id)
+    bool contains(Id id) const;
+
+    /// @replaces libdnf:sack/packageset.hpp:method:PackageSet.remove(Id id)
+    void remove(Id id);
+
+    // SET OPERATIONS - Map
+
+    /// Union operator
+    ///
+    /// @replaces libdnf:sack/packageset.hpp:method:PackageSet.operator+=(const Map * other)
+    SolvMap & operator|=(const Map * other);
+
+    /// Difference operator
+    ///
+    /// @replaces libdnf:sack/packageset.hpp:method:PackageSet.operator-=(const Map * other)
+    SolvMap & operator-=(const Map * other);
+
+    /// Intersection operator
+    ///
+    /// @replaces libdnf:sack/packageset.hpp:method:PackageSet.operator/=(const Map * other)
+    SolvMap & operator&=(const Map * other);
+
+    // SET OPERATIONS - SolvMap
+
+    /// Union operator
+    SolvMap & operator|=(const SolvMap & other);
+
+    /// Difference operator
+    SolvMap & operator-=(const SolvMap & other);
+
+    /// Intersection operator
+    SolvMap & operator&=(const SolvMap & other);
+
+protected:
+    /// Check if `id` is in bitmap range.
+    /// Throws std::out_of_range
+    void check_id_in_bitmap_range(Id id) const;
+
+    // ITEM OPERATIONS - UNSAFE
+
+    /// Faster, but unsafe version of add() method that is doesn't check bitmap range
+    void add_unsafe(Id id);
+
+    /// Faster, but unsafe version of contains() method that is doesn't check bitmap range
+    bool contains_unsafe(Id id) const;
+
+    /// Faster, but unsafe version of remove() method that is doesn't check bitmap range
+    void remove_unsafe(Id id);
+
+private:
+    Map map;
+};
+
+
+inline SolvMap::SolvMap(int size) {
+    map_init(&map, size);
+}
+
+
+inline SolvMap::SolvMap(const Map * other) {
+    map_init_clone(&map, other);
+}
+
+
+inline SolvMap::SolvMap(const SolvMap & other) : SolvMap(other.get_map()) {}
+
+
+inline SolvMap::~SolvMap() {
+    map_free(&map);
+}
+
+
+inline void SolvMap::check_id_in_bitmap_range(Id id) const {
+    if (id < 0 || id >= (map.size << 3)) {
```

Comment:
Please could you add a comment why the move is required? Believe that the size in libsolv is store in bites.

Reply:
Added a comment explaining the bit shift.